### PR TITLE
[GISEL] Add G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR

### PR DESCRIPTION
**Pull request by michaelmaitland**
G_INSERT and G_EXTRACT are not sufficient to use to represent both INSERT/EXTRACT on a subregister and INSERT/EXTRACT on a vector.

We would like to be able to INSERT/EXTRACT on vectors in cases that INSERT/EXTRACT on vector subregisters are not sufficient, so we add these opcodes.

I tried to do a patch where we treated G_EXTRACT as both G_EXTRACT_SUBVECTOR and G_EXTRACT_SUBREG, but ran into an infinite loop at this [point](https://github.com/llvm/llvm-project/blob/8b5b294ec2cf876bc5eb5bd5fcb56ef487e36d60/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L9932) in the SDAG equivalent code.

### CODE REVIEW

**Comment by topperc** [comment]
I just read the description of G_INSERT and G_EXTRACT. They sound exactly like G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR. They don't use a subreg index.

**Comment by topperc** [comment]
> I just read the description of G_INSERT and G_EXTRACT. They sound exactly like G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR. They don't use a subreg index.

Nevermind, I see now it specifically says "bit-index"

**Comment by arsenm** [comment]
> I just read the description of G_INSERT and G_EXTRACT. They sound exactly like G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR. They don't use a subreg index.

We should also get rid of G_INSERT/G_EXTRACT

File: llvm/test/MachineVerifier/test_g_extract_subvector.mir [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# RUN: not --crash llc -o - -mtriple=arm64 -run-pass=none -verify-machineinstrs %s 2>&1 | FileCheck %s
+# REQUIRES: aarch64-registered-target
```
Comment (arsenm) [inline llvm/test/MachineVerifier/test_g_extract_subvector.mir:2]: Don't really need to specify aarch64, you're not referencing anything target specific

**Comment by michaelmaitland** [comment]
> > I just read the description of G_INSERT and G_EXTRACT. They sound exactly like G_INSERT_SUBVECTOR and G_EXTRACT_SUBVECTOR. They don't use a subreg index.
> 
> We should also get rid of G_INSERT/G_EXTRACT

I think we need G_EXTRACT as the equivalent of EXTRACT_SUBREG in addition to G_INSERT_SUBVECTOR

File: llvm/test/MachineVerifier/test_g_extract_subvector.mir (hunk 1 above)
Comment (michaelmaitland) [reply llvm/test/MachineVerifier/test_g_extract_subvector.mir:2]: All tests in this directory have this require. Let’s remove them together in a single follow up patch?

Comment (michaelmaitland) [reply llvm/test/MachineVerifier/test_g_extract_subvector.mir:2]: I will remove this one here though!

Comment (michaelmaitland) [reply llvm/test/MachineVerifier/test_g_extract_subvector.mir]: Updated.

**Comment by arsenm** [comment]
> I think we need G_EXTRACT as the equivalent of EXTRACT_SUBREG in addition to G_INSERT_SUBVECTOR

Yes, it is the equivalent of EXTRACT_SUBREG, which is why it's problematic. It's essentially unlegalizable since it's too unrestricted

**Review by arsenm — APPROVED** [review]

File: llvm/test/MachineVerifier/test_g_extract_subvector.mir (hunk 1 above)
Comment (amilendra) [reply llvm/test/MachineVerifier/test_g_extract_subvector.mir]: > All tests in this directory have this require. Let’s remove them together in a single follow up patch?
I will remove this one here though!

These new tests fail cleanly without expectedly crashing on builds configured for selected targets (e.g. `-DLLVM_TARGETS_TO_BUILD=ARM`) without a `LLVM_DEFAULT_TARGET_TRIPLE`. 
```
bin/llc: error: unable to get target for '', see --version and --triple.
```
I think that is why tests in this directory have the `REQUIRES` along with a valid triple; either in command line (`-mtriple=aarch64`) or in the test file itself (`target triple = "aarch64--"`).

**Pull request merged**